airflow/metrics/otel_logger.py

```diff
@@ -0,0 +1,234 @@
+        self.meter = meter
+        self.map = {}
+
+    def clear(self) -> None:
+        self.map.clear()
+
+    def _create_counter(self, name):
+        """Creates a new counter or up_down_counter for the provided name."""
+        otel_safe_name = name[:OTEL_NAME_MAX_LENGTH]
+        if name != otel_safe_name:
+            warnings.warn(
+                f"Metric name `{name}` exceeds OpenTelemetry's name length limit of "
+                f"{OTEL_NAME_MAX_LENGTH} characters and will be truncated to `{otel_safe_name}`."
+            )
+
+        if _is_up_down_counter(name):
+            counter = self.meter.create_up_down_counter(name=otel_safe_name)
+        else:
+            counter = self.meter.create_counter(name=otel_safe_name)
+
+        print(f"--> created {otel_safe_name} as type: {str(type(counter)).split('.')[-1][:-2]}")
+        return counter
+
+    def get_counter(self, name: str, attributes: Attributes = None):
+        """
+        Returns the value of the counter; creates a new one if it does not exist.
+
+        :param name: The name of the counter to fetch or create.
+        :param attributes:  Counter attributes, used to generate a unique key to store the counter.
+        """
+        key: str = name + str(attributes)
+        if key in self.map.keys():
+            return self.map[key]
+        else:
+            new_counter = self._create_counter(name)
+            self.map[key] = new_counter
+            return new_counter
+
+    def del_counter(self, name: str, attributes: Attributes = None) -> None:
+        """
+        Deletes a counter.
+
+        :param name: The name of the counter to fetch or create.
+        :param attributes: Counter attributes which were used to generate a unique key to store the counter.
+        """
+        key: str = name + str(attributes)
+        if key in self.map.keys():
+            del self.map[key]
+
+
+def get_otel_logger(cls) -> SafeOtelLogger:
+    """Get Otel logger"""
+    host = conf.get("metrics", "otel_host")  # ex: "breeze-otel-collector"
+    port = conf.getint("metrics", "otel_port")  # ex: 4318
+    prefix = conf.get("metrics", "otel_prefix")  # ex: "airflow"
+    interval = conf.getint("metrics", "otel_interval_milliseconds")  # ex: 30000
+
+    allow_list = conf.get("metrics", "metrics_allow_list", fallback=None)
+    allow_list_validator = AllowListValidator(allow_list)
+
+    resource = Resource(attributes={SERVICE_NAME: "Airflow"})
+    # TODO:  figure out https instead of http ??
+    endpoint = f"http://{host}:{port}/v1/metrics"
+
+    print(f"[Metric Exporter] Connecting to OTLP at ---> {endpoint}")
+    readers = [
+        PeriodicExportingMetricReader(
+            OTLPMetricExporter(
+                endpoint=endpoint,
+                headers={"Content-Type": "application/json"},
+            ),
+            export_interval_millis=interval,
+        )
+    ]
+
+    # TODO:  remove this console exporter before launch??  It's handy for debugging maybe?
```

Comment:
Happy to hear opinions on this.  ConsoleMetricExporter prints every metric to the terminal instead of (or in addition to ) emitting them out to a Collector.  I found it very handy for debugging but likely less useful once this is all set up and working.  I intend to remove this, but open to leaving it and plumbing that debug flag through if others think it's worth keeping handy.

Reply:
If you are going to keep it, you should add a config option for it so that it can be enabled without modifying source code (like changing log level). But it's also fine to drop it, since it likely won't be used much.

Reply:
Added in https://github.com/apache/airflow/pull/30873/commits/9bf93d3b08de83c253b317142e746f7b5e469cca